src/main/kotlin/org/elaastic/questions/player/components/dashboard/DashboardModelFactory.kt

```diff
@@ -0,0 +1,232 @@
+
+/**
+ * Factory to build the [DashboardModel].
+ *
+ * @see DashboardModel
+ */
+@Service
+object DashboardModelFactory {
+
+    fun build(
+        sequence: Sequence,
+        previousSequence: Sequence?,
+        nextSequence: Sequence?,
+        attendees: List<LearnerAssignment>,
+        responses: List<Response>,
+        openedPane: String,
+        evaluationCountByUser: Map<LearnerAssignment, Long>
+    ): DashboardModel {
+
+        val learnerStepsModel: StepsModel = StepsModelFactory.buildForTeacher(sequence)
+
+        val sequenceMonitoringModel: SequenceMonitoringModel = SequenceMonitoringModel(
+            sequence.executionContext,
+            convertPhaseState(learnerStepsModel.responseSubmissionState),
+            convertPhaseState(learnerStepsModel.evaluationState),
+            convertPhaseState(learnerStepsModel.readState)
+        )
+
+        val learners: MutableList<LearnerMonitoringModel> = mutableListOf()
+
+        attendees.forEach {
+            learners.add(
+                LearnerMonitoringModel(
+                    it.id!!,
+                    it.learner.getFullname(),
+                    getAttendeeStateOnResponsePhase(it, responses, sequenceMonitoringModel.phase1State),
+                    getAttendeeStateOnEvaluationPhase(
+                        it,
+                        sequence,
+                        sequenceMonitoringModel.phase2State,
+                        evaluationCountByUser
+                    ),
+                    getAttendeeStateOnReadPhase(
+                        sequence,
+                        sequenceMonitoringModel.phase3State,
+                        it,
+                        sequenceMonitoringModel.phase2State,
+                        evaluationCountByUser
+                    ),
+                    sequenceMonitoringModel
+                )
+            )
+        }
+
+        sequenceMonitoringModel.setLearners(learners)
+
+        return DashboardModel(
+            sequence,
+            openedPane,
+            previousSequence?.id,
+            nextSequence?.id,
+            sequenceMonitoringModel
+        )
+    }
+
+    /**
+     * Get the LearnerStateOnPhase of the attendee on the Response phase.
+     *
+     * @param attendee the attendee
+     * @param attendeeResponses the list of responses of the attendee
+     * @param responsePhaseState the state of the response phase
+     * @return the state of the attendee on the response phase
+     * @see LearnerStateOnPhase
+     */
+    private fun getAttendeeStateOnResponsePhase(
+        attendee: LearnerAssignment,
+        attendeeResponses: List<Response>,
+        responsePhaseState: DashboardPhaseState
+    ): LearnerStateOnPhase {
+
+        val hasResponseForPhase: Boolean = attendeeResponses.count {
+            it.attempt == 1
+                    && it.interaction.interactionType == InteractionType.ResponseSubmission
+                    && it.learner == attendee.learner
+        } == 1
```

Comment:
This code must be extracted in a method so that it is called only if needed.

Reply:
Fix in 1ca8297ac3805803f772b32b0f8d821aec571817

Reply:
Real fix in 555dbda9d568bc4571f2d5ece71d17e02ac53aed